presto-hive/src/main/java/com/facebook/presto/twitter/hive/thrift/ThriftGenericRow.java

```diff
@@ -0,0 +1,228 @@
+    private byte[] buf = null;
+    private int off = 0;
+    private int len = 0;
+
+    public ThriftGenericRow() {}
+
+    public ThriftGenericRow(Map<Short, Object> values)
+    {
+        this.values.putAll(values);
+    }
+
+    public class Fields implements TFieldIdEnum
+    {
+        private final short thriftId;
+        private final String fieldName;
+
+        Fields(short thriftId, String fieldName)
+        {
+            this.thriftId = thriftId;
+            this.fieldName = fieldName;
+        }
+
+        public short getThriftFieldId()
+        {
+            return thriftId;
+        }
+
+        public String getFieldName()
+        {
+            return fieldName;
+        }
+    }
+
+    public void read(TProtocol iprot) throws TException
+    {
+        TTransport trans = iprot.getTransport();
+        buf = trans.getBuffer();
+        off = trans.getBufferPosition();
+        TProtocolUtil.skip(iprot, TType.STRUCT);
+        len = trans.getBufferPosition() - off;
+    }
+
+    public void parse() throws TException
+    {
+        parse(null);
+    }
+
+    public void parse(short[] thriftIds) throws TException
+    {
+        Set<Short> idSet = thriftIds == null ? null : new HashSet(Arrays.asList(ArrayUtils.toObject(thriftIds)));
+        TMemoryInputTransport trans = new TMemoryInputTransport(buf, off, len);
+        TBinaryProtocol iprot = new TBinaryProtocol(trans);
+        TField field;
+        iprot.readStructBegin();
+        while (true) {
+            field = iprot.readFieldBegin();
+            if (field.type == TType.STOP) {
+                break;
+            }
+            if (idSet != null && !idSet.remove(Short.valueOf(field.id))) {
+                TProtocolUtil.skip(iprot, field.type);
+            }
+            else {
+                values.put(field.id, readElem(iprot, field.type));
+            }
+            iprot.readFieldEnd();
+        }
+        iprot.readStructEnd();
+    }
+
+    private Object readElem(TProtocol iprot, byte type) throws TException
+    {
+        switch (type) {
+            case TType.BOOL:
+                return iprot.readBool();
+            case TType.BYTE:
+                return iprot.readByte();
+            case TType.I16:
+                return iprot.readI16();
+            case TType.ENUM:
+            case TType.I32:
+                return iprot.readI32();
+            case TType.I64:
+                return iprot.readI64();
+            case TType.DOUBLE:
+                return iprot.readDouble();
+            case TType.STRING:
+                return iprot.readString();
+            case TType.STRUCT:
+                return readStruct(iprot);
+            case TType.LIST:
+                return readList(iprot);
+            case TType.SET:
+                return readSet(iprot);
+            case TType.MAP:
+                return readMap(iprot);
+            default:
+                TProtocolUtil.skip(iprot, type);
+        }
+        return null;
```

Comment:
Can we move this into the default clause to make it explicit? Also what does returning null mean? Could we instead throw a TypeNotFoundException or something like that? Typically returning null is an anti-pattern.